### DISCUSSION TITLE
Filter permission must have a matching alias defined

### DISCUSSION THREAD

**Discussion by bedrijfsportaal** [codeigniter4/shield#600] [Q&A] [answered]
Hi,

I'm writing page permissions for my users, this is defined in `App\Config\Filters.php` when accessing the page with the correct permissions it shows `filter permission:users.manage must have a matching alias defined.`.

**The alias and filter**
```php
    public $aliases = [
        'permission' => \CodeIgniter\Shield\Filters\PermissionFilter::class,
    ];

    public $filters = [
        'permission:users.manage' => [
            'before' => ['users/profile']
        ]
    ];
```

Is there something missing or not understand from this point, i'm lost right now..

**Comment by datamweb** [comment]
Hello, please run the following command first.
```console
composer require codeigniter4/shield:dev-develop
```
There is **no need** to:
```console
    public $aliases = [
        'permission' => \CodeIgniter\Shield\Filters\PermissionFilter::class,
    ];

```

just used :
```console
    public $filters = [
        'permission:users.manage' => [
            'before' => ['users/profile']
        ]
    ];
```

**Reply by bedrijfsportaal** [reply]
Hi,

Thank you for the response, i'm already on `"codeigniter4/shield": "dev-develop"` this doesn't change the issue.

**Reply by datamweb** [reply]
Run the following command:
```console
php spark env development
```
screen shot ?

And also a screenshot of the path files `vendor\codeigniter4\shield\src\Filters`.

**Reply by bedrijfsportaal** [reply]
Thank you for the response, i'm on `development` mode.

**Screenshot vendor**
<a href="https://ibb.co/wpvMrc1"><img src="https://i.ibb.co/wpvMrc1/shield-Path.png" alt="shield-Path" border="0"></a>

**Reply by datamweb** [reply]
We **do not have** `users.manage` by default. You can create this or use existing ones.

https://github.com/codeigniter4/shield/blob/2e582d9f805a2f94ec9c097a3c06ad20a8a45e03/src/Config/AuthGroups.php#L65-L73

**Reply by bedrijfsportaal** [reply]
The permission has been set on the permission array, should this go in Settings() as well or is the file good enough? 
Because through `auth()->user()->hasPermission('users.manage')` does work as expected.

**edit**
Tested with default `users.create` it has the same issue.

**Reply by datamweb** [reply]
If you need to edit, add or delete the permissions through UI(admin panel), it is better to save them in the Settings table, not in this file `app/Config/AuthGroups.php`.

It depends on you.

**Reply by bedrijfsportaal** [reply]
It's not required to be managed, the permissions are added by me in `App\Config\AuthGroups.php` but the permissions don't work on the `App\Config\Filters.php` keep getting the Alias message also on default permissions.

**Reply by datamweb** [reply]
OK, I see that you have a problem.
Please post the minimum code for review, including the controller, view, routes file, filter, and anything else I can reproduce your problem.

**Reply by bedrijfsportaal** [reply]
Basicly default, it has nothing special as i'm just starting at this point.

**Created**
* Controller `UserController.php` with default method `index()`
* Added `permission:users.create` in the `$filters` on the before (as shown above).
* In the table `auth_permissions_users` i've added user ID 1 with permission `users.create`.

**It works when**
You add the permission inside __construct like: `auth()->user()->hasPermission('users.create')` and throw exception if not allowed.

**It does not work on**
Adding the permission on `$filters` the error thrown (in my language) is: filter `permission:users.create` does not have an alias defined.

**Reply by datamweb** [reply]
@bedrijfsportaal Oh, thanks, I noticed that. Please use the following way now.
`app\Config\Routes.php`:

```php
$routes->get('users/profile', 'UserController::index', ['filter' => 'permission:admin.access']);
```
It seems that our documents need to be corrected.

**Comment by datamweb** [comment, marked as answer]
@kenjis please see here.

In documents CI, there is no explanation regarding the use of arguments in method [filters](https://codeigniter.com/user_guide/incoming/filters.html?highlight=filter#id10). Therefore, it seems, it is not possible to use arguments in this method.

According to the Shield docs, it should be updated.
https://github.com/codeigniter4/shield/blob/develop/docs/authorization.md#authorizing-via-filters

**Reply by kenjis** [reply]
Yes, `$filters` cannot have arguments.
The current Shield docs are not correct.

**Reply by bedrijfsportaal** [reply]
@kenjis & @datamweb  
That would explain what is going on, i'll change my work to setup on routes!

Thank you for the support on this matter!

**Reply by kenjis** [reply]
@datamweb 
> According to the Shield docs, it should be updated.
https://github.com/codeigniter4/shield/blob/develop/docs/authorization.md#authorizing-via-filters

I will take care of it.

Done. #606

**Reply by kenjis** [reply]
I sent a PR for `$filters` arguments.
See https://github.com/codeigniter4/CodeIgniter4/pull/7159
